### PR TITLE
Setup simple API for creating and retrieving plans, timeframes, and segments

### LINKED ISSUE #24: Plan Builder: Create API for creating and fetching plans

### PR DESCRIPTION
**Pull request by iangmhill**
Resolves #24 for the time being. Given time constraints, this is a rough cut version of the API. We will need real tests for it in the future.

### CODE REVIEW

File: client/app/components/Viewer.jsx [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import PropTypes from 'prop-types';
 import React from 'react';
 
-import MapContainer from '../containers/MapContainer';
-import PlanFormContainer from '../containers/PlanFormContainer';
+import MapContainer from './../containers/MapContainer';
```
Comment (ericyoungberg) [inline client/app/components/Viewer.jsx]: Isn't this a redundancy?

File: config/routes.rb [hunk 2]
```diff
@@ -15,5 +15,10 @@
         get '/' => 'node#show'
       end
     end
+    scope '/plan' do
+      get '/' => 'plan#index'
+      post '/' => 'plan#create'
+      get '/:id' => 'plan#show'
+    end
```
Comment (ericyoungberg) [inline config/routes.rb]: The way you have setup the _routes.rb_ file doesn't adhere to the Rails philosophy. What you are doing works but it is effectively what Rails will do for you should you configure it the way they define in the docs. This is a nice implementation of an Express-like approach in Rails though! 

- [CRUD Verbs and Actions](http://guides.rubyonrails.org/routing.html#crud-verbs-and-actions)
- [Controller Namespaces and Routing](http://guides.rubyonrails.org/routing.html#controller-namespaces-and-routing)
- [Restricting the Routes Created](http://guides.rubyonrails.org/routing.html#restricting-the-routes-created)

File: db/migrate/20180329141702_add_plan_and_segment_tables.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    create_table :timeframes do |t|
+      t.belongs_to :plan, index: true
+      t.date :start
+      t.date :end
+      t.timestamps
+    end
+    create_table :segments do |t|
+      t.belongs_to :timeframe, index: true
+      t.belongs_to :road, index: true
+      t.boolean :is_whole_road
```
Comment (ericyoungberg) [inline db/migrate/20180329141702_add_plan_and_segment_tables.rb]: This is a philosophical comment more than something that actually needs to be changed: 

As programmers, we are constantly checking to see if the state of something is true. Truthiness can often be found in most dynamic languages in just about every type as long as there is a value assigned. This being the case, truthiness should be used as an active state, not the default. Since our interface has the default state being a whole road, the default for `:is_whole_road` is `true`. It is more sound to have the column named `:is_segment` so we can adhere to our default state when checking for truthiness.

Comment (ericyoungberg) [reply db/migrate/20180329141702_add_plan_and_segment_tables.rb]: Additionally (this is something that probably should be included if we are being defensive and thorough), the nature of booleans is that there are only two possible values. If Postgres allows `NULL` in boolean columns, `typeof null !== 'boolean'`. Since your default state assumes `true` based on the reason in my comment above, `null == false` which would will trigger conditionals to not be triggered that we may want to.  

**tl;dr** Always set default values for boolean columns based upon the default state that the column represents.

File: client/app/components/Viewer.jsx (hunk 1 above)
Comment (iangmhill) [reply client/app/components/Viewer.jsx]: Generally, yes. I was trying to debug what I thought was an import error in Webpack at one point, and I wanted to give this a try. It's just a leftover remnant of that test that can now be removed.

File: config/routes.rb (hunk 2 above)
Comment (iangmhill) [reply config/routes.rb]: This black boxing is very hard to get used to. I'll fix that in the next version of this PR.

File: db/migrate/20180329141702_add_plan_and_segment_tables.rb (hunk 3 above)
Comment (iangmhill) [reply db/migrate/20180329141702_add_plan_and_segment_tables.rb]: I was attempting to match the way that we asked the question in the UI where the TRUE position indicates that the user is talking about the whole road. But I see your point and will change this.

**Comment by iangmhill** [comment]
Test curl:

`curl -X 'POST' http://localhost:3000/api/plan -H "Content-Type: application/json" -d '{
  "plan": {
    "name": "Test Plan",
    "type": "PAVING",
    "city": "AYER",
    "published": true,
    "timeframes": [{
      "start": "2018-03-29T19:37:48.454Z",
      "end": "2018-03-29T19:37:48.454Z",
      "segments": [{
        "road_id": 22,
        "is_segment": true,
        "is_orig_type_address": false,
        "orig": 4,
        "is_dest_type_address": false,
        "dest": 5,
        "nodes": [1,2,3],
        "custom_nodes": ""
      }]
    }]
  }
}' > tmp.html`

File: app/controllers/api/plan_controller.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      planHash[:timeframes_attributes].each.with_index do |timeframe, index|
+        planHash[:timeframes_attributes][index][:segments_attributes] = planHash[:timeframes_attributes][index][:segments]
+        planHash[:timeframes_attributes][index].delete(:segments)
+      end
+      planHash.permit!
+      plan = Plan.create!(planHash)
+      respond_to do |format|
+        format.json {
+          render json: {
+            success: true,
```
Comment (mzagaja) [inline app/controllers/api/plan_controller.rb]: Is returning `success: true` in JSON redundant to the HTTP status code response?
Does `create!` effectively return an error to the client with a relevant HTTP status code? Here is the pattern I typically use: https://github.com/MAPC/massbuilds/blob/develop/rails/app/controllers/developments_controller.rb#L50-L56

**Comment by iangmhill** [comment]
All of these comments have been addressed in the latest commit.
